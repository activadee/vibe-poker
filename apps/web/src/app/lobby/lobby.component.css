/*
  Legacy lobby styles removed.
  The lobby now uses Tailwind utilities and the design system directives
  (appUiButton, appUiInput, appUiCheckbox, app-ui-card) for layout and styling.
*/
