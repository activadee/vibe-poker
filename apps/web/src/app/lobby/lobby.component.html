<div class="lobby">
  <h1>Vibe Kanban</h1>

  <section aria-labelledby="create-title">
    <h2 id="create-title">Create a room</h2>
    <p>Enter your display name to start a new room.</p>
    <form (ngSubmit)="createRoom()" autocomplete="on">
      <label for="displayName">Display Name</label>
      <input
        id="displayName"
        name="displayName"
        placeholder="e.g. Alice"
        [(ngModel)]="name"
        minlength="3"
        maxlength="30"
        required
        autocomplete="name"
        aria-describedby="nameHelp"
      />
      <small id="nameHelp" class="muted">3–30 characters</small>
      <div class="actions">
        <button type="submit" [disabled]="loading() || !nameValid()">Create Room</button>
      </div>
    </form>
  </section>

  <section aria-labelledby="join-title">
    <h2 id="join-title">Join a room</h2>
    <p>Paste a room code or full link to join.</p>
    <form (ngSubmit)="joinRoom()">
      <label for="roomCode">Room Code or Link</label>
      <input
        id="roomCode"
        name="roomCode"
        placeholder="ABCD-1234 or https://…/r/ABCD-1234"
        [(ngModel)]="roomCode"
        aria-describedby="roomHelp"
      />
      <small id="roomHelp" class="muted">Join is enabled when code is present</small>
      <label class="check">
        <input type="checkbox" [(ngModel)]="joinAsObserver" name="joinAsObserver" />
        Join as observer (cannot vote)
      </label>
      <div class="actions">
        <button type="submit" [disabled]="!canJoin()">Join Room</button>
      </div>
    </form>
  </section>

  <p class="error" *ngIf="error()" role="alert">{{ error() }}</p>
</div>
