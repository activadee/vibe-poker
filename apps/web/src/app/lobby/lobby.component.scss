.lobby {
  max-width: 480px;
  margin: 48px auto;
  padding: 16px;
}

section { margin-bottom: 24px; }

label {
  display: block;
  font-weight: 600;
}

input {
  width: 100%;
  padding: 8px;
  margin: 8px 0;
}

button {
  padding: 8px 12px;
}

.muted { color: #666; font-size: 12px; }

.actions { margin-top: 8px; }

.error {
  color: #b00020;
}
