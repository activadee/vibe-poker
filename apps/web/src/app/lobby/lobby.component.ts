import { Component, inject, signal } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { HttpClient } from '@angular/common/http';
import { Router } from '@angular/router';
import { CreateRoomResponse, CreateRoomRequest } from '@scrum-poker/shared-types';
import { firstValueFrom } from 'rxjs';
import { UiButtonDirective } from '../ui/button.directive';
import { UiInputDirective } from '../ui/input.directive';
import { UiCheckboxDirective } from '../ui/checkbox.directive';
import { UiCardComponent } from '../ui/card/card.component';

@Component({
  selector: 'app-lobby',
  standalone: true,
  imports: [CommonModule, FormsModule, UiButtonDirective, UiInputDirective, UiCheckboxDirective, UiCardComponent],
  templateUrl: './lobby.component.html',
  styleUrls: ['./lobby.component.scss'],
})
export class LobbyComponent {
  private readonly http = inject(HttpClient);
  private readonly router = inject(Router);

  name = '';
  roomCode = '';
  joinAsObserver = false;
  loading = signal(false);
  error = signal('');

  nameValid() {
    const v = (this.name || '').trim();
    return v.length >= 3 && v.length <= 30;
  }
  canJoin() {
    return (this.roomCode || '').trim().length > 0;
  }

  async createRoom() {
    const hostName = this.name.trim();
    if (!this.nameValid()) {
      this.error.set('Display name must be 3â€“30 characters');
      return;
    }
    this.error.set('');
    this.loading.set(true);
    try {
      const res = await firstValueFrom(
        this.http.post<CreateRoomResponse>(
          '/api/rooms',
          { hostName } satisfies CreateRoomRequest
        )
      );
      if (res?.id) {
        try {
          localStorage.setItem('displayName', hostName);
        } catch (err) {
          console.warn('localStorage unavailable', err);
        }
        // Warm the Room chunk before navigating for a snappier transition
        void import('../room/room.component');
        // Navigate with a flag so the room view can auto-join the host
        this.router.navigate(['/r', res.id], { queryParams: { host: '1' } });
      } else {
        this.error.set('Unexpected response from server');
      }
    } catch (err) {
      // Keep user-facing message simple while still surfacing errors during development
      console.error('Failed to create room', err);
      this.error.set('Failed to create room');
    } finally {
      this.loading.set(false);
    }
  }

  joinRoom() {
    const input = (this.roomCode || '').trim();
    if (!input) return;
    // Save name if valid (non-blocking for join per AC)
    const hostName = (this.name || '').trim();
    if (hostName.length >= 3 && hostName.length <= 30) {
      try { localStorage.setItem('displayName', hostName); } catch (err) { console.warn('localStorage unavailable', err); }
    }
    const roomId = LobbyComponent.extractRoomId(input);
    if (!roomId) {
      this.error.set('Please enter a valid room code or link');
      return;
    }
    this.error.set('');
    const role = this.joinAsObserver ? 'observer' : 'player';
    // Warm the Room chunk before navigating for a snappier transition
    void import('../room/room.component');
    this.router.navigate(['/r', roomId], { queryParams: { role } });
  }

  static extractRoomId(input: string): string | null {
    const s = (input || '').trim();
    if (!s) return null;
    const rIndex = s.lastIndexOf('/r/');
    if (rIndex !== -1) {
      const after = s.substring(rIndex + 3);
      const id = after.split(/[?/#]/)[0];
      return id || null;
    }
    // Fallback: treat as code; normalize whitespace
    return s || null;
  }
}
