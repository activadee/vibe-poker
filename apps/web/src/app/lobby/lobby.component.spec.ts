import { TestBed } from '@angular/core/testing';
import { LobbyComponent } from './lobby.component';
import { HttpClientTestingModule, HttpTestingController } from '@angular/common/http/testing';
import { Router } from '@angular/router';

describe('LobbyComponent', () => {
  let httpMock: HttpTestingController;
  let navigateSpy: jest.Mock;

  beforeEach(() => {
    navigateSpy = jest.fn();
    TestBed.configureTestingModule({
      imports: [LobbyComponent, HttpClientTestingModule],
      providers: [
        { provide: Router, useValue: { navigate: navigateSpy } },
      ],
    });
    httpMock = TestBed.inject(HttpTestingController);
  });

  it('creates a room and navigates to /r/:id', async () => {
    const fixture = TestBed.createComponent(LobbyComponent);
    const comp = fixture.componentInstance;
    comp.name = 'Alice';

    const promise = comp.createRoom();

    const req = httpMock.expectOne('/api/rooms');
    expect(req.request.method).toBe('POST');
    expect(req.request.body).toEqual({ hostName: 'Alice' });
    req.flush({ id: 'ABCD-1234', expiresAt: Date.now() + 86400000 });

    await promise;

    expect(navigateSpy).toHaveBeenCalledWith(['/r', 'ABCD-1234']);
  });
});
