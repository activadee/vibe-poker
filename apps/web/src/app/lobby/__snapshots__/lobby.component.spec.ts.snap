// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LobbyComponent renders initial state (snapshot) 1`] = `
<div
  id="root1"
>
  <div
    class="lobby"
  >
    <h1>
      Vibe Kanban
    </h1>
    <section
      aria-labelledby="create-title"
    >
      <h2
        id="create-title"
      >
        Create a room
      </h2>
      <p>
        Enter your display name to start a new room.
      </p>
      <form
        autocomplete="on"
        class="ng-untouched ng-pristine ng-valid"
        novalidate=""
      >
        <label
          for="displayName"
        >
          Display Name
        </label>
        <input
          aria-describedby="nameHelp"
          autocomplete="name"
          class="ng-untouched ng-pristine ng-valid"
          id="displayName"
          maxlength="30"
          minlength="3"
          name="displayName"
          placeholder="e.g. Alice"
          required=""
        />
        <small
          class="muted"
          id="nameHelp"
        >
          3–30 characters
        </small>
        <div
          class="actions"
        >
          <button
            disabled=""
            type="submit"
          >
            Create Room
          </button>
        </div>
      </form>
    </section>
    <section
      aria-labelledby="join-title"
    >
      <h2
        id="join-title"
      >
        Join a room
      </h2>
      <p>
        Paste a room code or full link to join.
      </p>
      <form
        class="ng-untouched ng-pristine ng-valid"
        novalidate=""
      >
        <label
          for="roomCode"
        >
          Room Code or Link
        </label>
        <input
          aria-describedby="roomHelp"
          class="ng-untouched ng-pristine ng-valid"
          id="roomCode"
          name="roomCode"
          placeholder="ABCD-1234 or https://…/r/ABCD-1234"
        />
        <small
          class="muted"
          id="roomHelp"
        >
          Join is enabled when code is present
        </small>
        <div
          class="actions"
        >
          <button
            disabled=""
            type="submit"
          >
            Join Room
          </button>
        </div>
      </form>
    </section>
    
  </div>
</div>
`;
